qiita_ware/ebi.py

```diff
@@ -587,6 +595,20 @@ def generate_run_xml(self):
                 'checksum': md5}
             )
 
+            # inserting reverse
+            if self.per_sample_FASTQ_reverse:
+                file_path = self.sample_demux_fps[
```

Comment:
I think the next 10 lines of code are identical to the 10 lines starting on line 586 with the exception of the suffix used? What about

```python
def _add_file_subelement(self, add_file, sample_name, is_forward):
    if is_forward:
        suffix = self.FWD_READ_SUFFIX
    else:
        suffix = self.REV_READ_SUFFIX

    file_path = self.sample_demux_fps[sample_name] + suffix
    with open(file_path) as fp:
        md5 = safe_md5(fp).hexdigest()
    
    file_details = {'filetype': file_type,
                    'quality_scoring_system': 'phred',
                    'checksum_method': 'MD5',
                    'checksum': md5,
                    'filename': join(self.ebi_dir, basename(file_path))}
   
    add_file(file_details)
```

...and then in `generate_run_xml`, do:

```python
    add_file = functools.partial(ET.SubElement, files, 'FILE')
    self._add_file_subelement(add_file, sample_name, is_forward=False)
    if self.per_sample_FASTQ_reverse:
        _add_file_subelement(add_file, sample_name, is_forward=False)
```

Testing of `_add_file_subelement` can be done with a mocked out ET as well so no need to spin up complex ET logic.

Reply:
Like it! ... but what about making _add_file_subelement a function that only exists in generate_run_xml? Which is the only place where is being used ...